### PR TITLE
C++ time & timeless support

### LINKED ISSUE #3394: C++ implement logging with timestamps
and without default timeline! Right now all one can do is to log in `log_tick` & `log_time`

### PR DESCRIPTION
**Pull request by Wumpf**
### What

* Fixes #3394

Introduces time methods to C/C++, timeless logging and ports the clock example.

Originally, I planned to expose a Timeline struct type to C but this actually created more surface for errors to occur, so I instead expose the three time setter for seconds/nanos/sequences separately, making the code much easier to reason with.

image

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [demo.rerun.io](https://demo.rerun.io/pr/3838) (if applicable)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/3838)
- [Docs preview](https://rerun.io/preview/81bd132335ca725ae0c2a1fb2c42156f5b1986d3/docs) 
- [Examples preview](https://rerun.io/preview/81bd132335ca725ae0c2a1fb2c42156f5b1986d3/examples) 
- [Recent benchmark results](https://ref.rerun.io/dev/bench/)
- [Wasm size tracking](https://ref.rerun.io/dev/sizes/)

### CODE REVIEW

File: examples/cpp/clock/main.cpp [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+#include <cmath>
+#include <rerun.hpp>
+
+#define TAU (M_PI * 2.0)
+
+void log_hand(
+    rerun::RecordingStream& rec, const char* name, int step, float angle, float length, float width,
+    uint8_t blue
+) {
+    rerun::datatypes::Vec3D pos(length * sinf(angle * TAU), length * cosf(angle * TAU), 0.0f);
```
Comment (emilk) [inline examples/cpp/clock/main.cpp:11]: `tip` would be a more explanatory name for this

File: crates/rerun_c/src/lib.rs [hunk 2]
```diff
@@ -301,17 +308,115 @@ pub extern "C" fn rr_recording_stream_save(
...
+    stream: CRecordingStream,
+    timeline_name: *const c_char,
+    sequence: *const i64,
+) -> Result<(), CError> {
+    let timeline = ptr::try_char_ptr_as_str(timeline_name, "timeline_name")?;
+    let sequence = if sequence.is_null() {
+        None
+    } else {
+        Some(unsafe { *sequence })
+    };
```
[the remark this answers is not in the document]
Comment (Wumpf) [reply crates/rerun_c/src/lib.rs:323]: ohhhh nice!

File: examples/cpp/clock/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,15 @@
+cmake_minimum_required(VERSION 3.16)
+
+# project(ClockExample)
```
[the remark this answers is not in the document]
Comment (Wumpf) [reply examples/cpp/clock/CMakeLists.txt:3]: same reason it is commented out in the minimal example: The project is still `rerun` but if you want to copy this out you need to copy this out and create a project here.
But let's remove this right now, we need to do more work in this area anyways and explain what to do

File: rerun_cpp/src/rerun/recording_stream.hpp [hunk 4]
```diff
@@ -120,6 +120,51 @@ namespace rerun {
...
+        /// Set the current time of the recording, for the current calling thread.
+        ///
+        /// Used for all subsequent logging performed from this same thread, until the next call
+        /// to one of the time setting methods.
+        ///
+        /// For example: `rec.set_time_nanos("sim_time", sim_time_nanos)`.
+        ///
+        /// You can remove a timeline again using `rec.set_time_nanos("sim_time", std::null_opt)`.
+        /// @see set_timepoint, set_time_sequence, set_time_seconds, reset_time,
+        void set_time_nanos(const char* timeline_name, std::optional<int64_t> nanos);
```
[the remark this answers is not in the document]
Comment (Wumpf) [reply rerun_cpp/src/rerun/recording_stream.hpp:157]: leaned heavily on what Rust does. But agreed this is much better here. Rolling out to C we also don't have to deal with pointers there then which means less crash potential!!

File: examples/cpp/clock/main.cpp [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    uint8_t c = static_cast<uint8_t>(angle * 255.0f);
+    rerun::components::Color color(255 - c, c, blue, std::max<uint8_t>(128, blue));
```
[the remark this answers is not in the document]
Comment (Wumpf) [reply examples/cpp/clock/main.cpp:13]: urgh. Not a fan of this style at all but you're right I think.
Would we then put curlies on all initializations everywhere if we put this in the style guide?

**Pull request merged**